### DISCUSSION TITLE
How to connect previously trained Spacy NER model to new empty Language pipeline?

### DISCUSSION THREAD

**Discussion by justmars** [explosion/spaCy#13148] [Help: Coding & Implementations] [answered]
I've been able to train and load a model using a custom script involving a `@Language` factory:

```py
# post training
_model = spacy.load("./models/statutes/model-best")
_model.pipe_names
# ['tok2vec', 'ner']
doc = _model('Section 13 of PD No. 1869 provides xxx')
for e in doc.ents:
  print(f"{e.text=} {e.label_=}")
# e.text='PD No. 1869' e.label_='STATUTE'
```

I'd like to load this model in a separate language object and it seems to work but no entities are detected:

```py
x = spacy.blank("en")
x.add_pipe("ner", name="statute_ner", source=_model)
x.pipe_names
# ['statute_ner']
doc = x('Section 13 of PD No. 1869 provides xxx')
doc.ents
# ()
```

Note this follows the convention described in #10674.

According to the docs:

> Important note for trained components
> When reusing components across pipelines, keep in mind that the vocabulary, vectors and model settings must match.

Initially I thought that since `tok2vec` wasn't in the blank  model, it would fail. So I tried loading the `en_core_web_sm` model which had it but it still failed to detect the entities from the pretrained `_model`. 

I think the vocabulary matches. Wehn running `x.to_disk`, the vocab found in the `_model` trained appears in `x`. So I'm guessing I can rule this out.

There was some discussion about loading files via `Language.factories` through an `__init__.py` file found in a [dated clip of the quite thorough Dr. WJB Mattingly / Python for Digital Humanities](https://youtu.be/AGM7lXfQECM?si=o3_L45Uab1Ji0fJa&t=745) but it seems like this has been eclipsed by spacy v3.

Hope for some guidance on what I'm doing wrong. Thank you!

**Comment by justmars** [comment, marked as answer]
Ok, upon investigating #10674, @adrianeboyd links to https://github.com/explosion/projects/tree/v3/tutorials/ner_double and I see a curious: `drug_nlp.replace_listeners("tok2vec", "ner", ["model.tok2vec"])`. Didn't know this was a necessary step so I tried it and it solves my problem. Thank you!

```py
_model = spacy.load("./models/statutes/model-best")
_model.pipe_names
# ['tok2vec', 'ner']
# adapted from solution
_model.replace_listeners("tok2vec", "ner", ["model.tok2vec"]) 
# reuse same code above
x = spacy.blank("en")
x.add_pipe("ner", name="statute_ner", source=_model)
x.pipe_names
# ['statute_ner']
doc = x('Section 13 of PD No. 1869 provides xxx')
for e in doc.ents:
  print(f"{e.text=} {e.label_=}")
# e.text='PD No. 1869' e.label_='STATUTE'
```

**Reply by rmitsch** [reply]
Happy you could resolve this, and thanks for posting the solution!
